cluster/tls_connection.go

```diff
@@ -0,0 +1,175 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/memberlist"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/alertmanager/cluster/clusterpb"
+)
+
+const (
+	version      = "v0.1.0"
+	uint32length = 4
+)
+
+// tlsConn wraps net.Conn with connection pooling data.
+type tlsConn struct {
+	mtx        sync.Mutex
+	connection net.Conn
+	live       bool
+}
+
+func dialTLSConn(addr string, timeout time.Duration, tlsConfig *tls.Config) (*tlsConn, error) {
+	dialer := &net.Dialer{Timeout: timeout}
+	conn, err := tls.DialWithDialer(dialer, network, addr, tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tlsConn{
+		connection: conn,
+		live:       true,
+	}, nil
+}
+
+func rcvTLSConn(conn net.Conn) *tlsConn {
+	return &tlsConn{
+		connection: conn,
+		live:       true,
+	}
+}
+
+// Write writes a byte array into the connection. It returns the number of bytes written and an error.
+func (conn *tlsConn) Write(b []byte) (int, error) {
+	conn.mtx.Lock()
+	defer conn.mtx.Unlock()
+	n, err := conn.connection.Write(b)
+
+	if err != nil {
+		conn.live = false
+	}
+	return n, err
+}
+
+func (conn *tlsConn) alive() bool {
+	conn.mtx.Lock()
+	defer conn.mtx.Unlock()
+	return conn.live
+}
+
+// writePacket writes all the bytes in one operation so no concurrent write happens in between.
+// It prefixes the message length.
+func (conn *tlsConn) writePacket(fromAddr string, b []byte) error {
+	msg, err := proto.Marshal(
+		&clusterpb.MemberlistMessage{
+			Version:  version,
+			Kind:     clusterpb.MemberlistMessage_PACKET,
+			FromAddr: fromAddr,
+			Msg:      b,
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal memeberlist packet message")
+	}
+	buf := make([]byte, uint32length, uint32length+len(msg))
+	binary.LittleEndian.PutUint32(buf, uint32(len(msg)))
+	_, err = conn.Write(append(buf, msg...))
+	return err
+}
+
+// writeStream simply signals that this is a stream connection by sending the connection type.
+func (conn *tlsConn) writeStream() error {
+	msg, err := proto.Marshal(
+		&clusterpb.MemberlistMessage{
+			Version: version,
+			Kind:    clusterpb.MemberlistMessage_STREAM,
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal memeberlist stream message")
+	}
+	buf := make([]byte, uint32length, uint32length+len(msg))
+	binary.LittleEndian.PutUint32(buf, uint32(len(msg)))
+	_, err = conn.Write(append(buf, msg...))
+	return err
+}
+
+// read returns a packet for packet connections or an error if there is one.
+// It returns nothing if the connection is meant to be streamed.
+func (conn *tlsConn) read() (*memberlist.Packet, error) {
+	if conn.connection == nil {
```

Comment:
my understanding was that `conn.mtx` is protecting connection, any particular reason why we're not using the lock to read the connection here?

Reply:
Good catch. I looked at it more and I think the lock was incidentally unnecessary because `read` is only called after `rcvTLSConn`, which creates a new connection object.

However, I'm adding a lock around the reading of the connection to make this safer to use.